msm/src/logup.rs

```diff
@@ -561,115 +573,120 @@ pub mod prover {
-                        for f_i in f.iter() {
-                            let Logup {
-                                numerator,
-                                table_id: _,
-                                value: _,
-                            } = &f_i[j as usize];
-                            row_acc += *numerator * denominators[denominator_index];
-                            denominator_index += 1;
-                            // We split in chunks of (MAX_SUPPORTED_DEGREE - 2)
-                            // We reset the accumulator for the current partial
-                            // sum after keeping it.
-                            if denominator_index % (MAX_SUPPORTED_DEGREE - 2) == 0 {
-                                partial_sums[partial_sum_idx].push(row_acc);
-                                row_acc = G::ScalarField::zero();
-                                partial_sum_idx += 1;
-                            }
-                        }
-                        if denominator_index % (MAX_SUPPORTED_DEGREE - 2) != 0 {
+                        // β + a_{i}
+                        let lookup_denominator = beta + combined_value;
+                        denominators.push(lookup_denominator);
+                    }
+                }
+                assert!(denominators.len() == n * domain.d1.size as usize);
+
+                ark_ff::fields::batch_inversion(&mut denominators);
+
+                // Evals is the sum on the individual columns for each row
+                let mut denominator_index = 0;
+
+                // We only need to add the numerator now
+                for j in 0..domain.d1.size {
+                    let mut partial_sum_idx = 0;
+                    let mut row_acc = G::ScalarField::zero();
+                    for (i, f_i) in f.iter().enumerate() {
+                        let Logup {
+                            numerator,
+                            table_id: _,
+                            value: _,
+                        } = &f_i[j as usize];
+                        row_acc += *numerator * denominators[denominator_index];
+                        denominator_index += 1;
+                        // We split in chunks of (MAX_SUPPORTED_DEGREE - 2)
+                        // We reset the accumulator for the current partial
+                        // sum after keeping it.
+                        if (i + 1) % (MAX_SUPPORTED_DEGREE - 2) == 0 {
                             partial_sums[partial_sum_idx].push(row_acc);
+                            row_acc = G::ScalarField::zero();
+                            partial_sum_idx += 1;
                         }
                     }
-                    partial_sums
-                })
-                .collect::<Vec<_>>();
-
-            let lookup_terms_evals: Vec<Vec<G::ScalarField>> =
-                lookup_terms_evals.into_iter().flatten().collect();
+                    if n % (MAX_SUPPORTED_DEGREE - 2) != 0 {
```

Comment:
```suggestion
                    // Whatever leftover in `row_acc` left in the end of the iteration, we write it into
                    // `partial_sums` too. This is only done in case `n % (MAX_SUPPORTED_DEGREE - 2) != 0`
                    // which means that the similar addition to `partial_sums` a few lines above won't be triggered.
                    // So we have this wrapping up call instead.
                    if n % (MAX_SUPPORTED_DEGREE - 2) != 0 {
```